README.md
# WebBench

Webbench是一个在linux下使用的非常简单的网站压测工具。它使用fork()模拟多个客户端同时访问我们设定的URL，测试网站在压力下工作的性能，最多可以模拟3万个并发连接去测试网站的负载能力。

这是我第一次改写一个C项目，虽然结果不太成功，但是也学会了不少东西，总结的东西没有多少，就那这个仓库当做结贴把。

我主要做的东西是，把webbench的一个500行的代码拆成了4个文件，更好的阅读和解耦。

分别是 socket.h, socket.cpp, bench.cpp(核心)，main.cpp(启动)，build_request.cpp构建请求头。

# 经验
1. 一次走一小步。
2. c语言和C++差别很大，光是字符串的api就够折磨人。 
3. 面向过程编程真的很难受。
4. 段错误和各种报错真非常 折磨人， 所以尽早学会自己调试是非常重要的。

# 总结
1. 虽然没有达到预期的目标，但是也算走了80%，勉强算成功，花了4天的时间，去解读这个项目

# 时间
2021-01-27 到2021-01-22，我珍贵的4天时间。
 
  
 

                     |


debian/changelog
webbench (1.5) unstable; urgency=low

  * allow building with both Gnu and BSD make

 -- Radim Kolar <hsn@home>  Fri, Jun 25 12:00:20 CEST 2004

webbench (1.4) unstable; urgency=low

  * check if url is not too long
  * report correct program version number
  * use yield() when waiting for test start
  * corrected error codes
  * check availability of test server first
  * do not abort test if first request_str failed
  * report when some childrens are dead.
  * use alarm, not time() for lower syscal use by bench
  * use mode 644 for installed doc
  * makefile cleaned for better freebsd ports integration

 -- Radim Kolar <hsn@home>  Thu, 15 Jan 2004 11:15:52 +0100

webbench (1.3) unstable; urgency=low

  * Build fixes for freeBSD
  * Default benchmark time 60 -> 30
  * generate tar with subdirectory
  * added to freeBSD ports collection

 -- Radim Kolar <hsn@home>  Mon, 12 Jan 2004 17:00:24 +0100

webbench (1.2) unstable; urgency=low

  * Only debian-related bugfixes
  * Updated Debian/rules
  * Adapted to fit new directory system
  * moved from debstd to dh_*

 -- Radim Kolar <hsn@home>  Fri, 18 Jan 2002 12:33:04 +0100

webbench (1.1) unstable; urgency=medium
 
  * Program debianized
  * added support for multiple methods (GET, HEAD, OPTIONS, TRACE)
  * added support for multiple HTTP versions (0.9 -- 1.1)
  * added long options
  * added multiple clients
  * wait for start of second before test
  * test time can be specified
  * better error checking when reading reply from server
  * FIX: tests was one second longer than expected

 -- Radim Kolar <hsn@home>  Thu, 16 Sep 1999 18:48:00 +0200

Local variables:
mode: debian-changelog
End:


main.cpp
//
// Created by zxzx on 2021/1/21.
//

#include <unistd.h>
#include <sys/param.h>
#include <rpc/types.h>
#include <getopt.h>
#include <strings.h>
#include <cstdio>
#include <cstdlib>

#include "socket.h"

bench_request bench_req;

static const struct option long_options[]=
{
		{"force",no_argument,& bench_req.force,1},
		{"reload",no_argument,& bench_req.force_reload,1},
		{"time",required_argument,NULL,'t'},
		{"help",no_argument,NULL,'?'},
		{"http09",no_argument,NULL,'9'},
		{"http10",no_argument,NULL,'1'},
		{"http11",no_argument,NULL,'2'},
		{"get",no_argument,& bench_req.method,METHOD_GET},
		{"head",no_argument,& bench_req.method,METHOD_HEAD},
		{"options",no_argument,& bench_req.method,METHOD_OPTIONS},
		{"trace",no_argument,& bench_req.method,METHOD_TRACE},
		{"version",no_argument,NULL,'V'},
		{"proxy",required_argument,NULL,'p'},
		{"clients",required_argument,NULL,'c'},
		{NULL,0,NULL,0}
};

//
void usage(void);
//// 外部的三个方法
void build_request(bench_request *req);
int bench(bench_request* req);
void f1(char* chars);
int main(int argc, char *argv[]){
	bench_req.proxyport = 80;
	bench_req.clients = 1;
	bench_req.method=METHOD_GET;
	if(argc==1)
	{
		usage();
		return 2;
	}

	int opt=0;
	int options_index=0;
	char* tmp;
	while((opt=getopt_long(argc,argv,"912Vfrt:p:c:?h",long_options,&options_index))!=EOF ){
		switch(opt){
			case 'c':
				bench_req.clients=atoi(optarg);
				break;
			case  0 :
				break;
			case 'f':
				bench_req.force=1;
				break;
			case 'r':
				bench_req.force_reload=1;
				break;
			case '9':
				bench_req.http10=0;
				break;
			case '1':
				bench_req.http10=1;
				break;
			case '2':
				bench_req.http10=2;
				break;
			case 'V':
				printf("%s\n", PROGRAM_VERSION);
				exit(0);
			case 't':
				bench_req.benchtime=atoi(optarg);
				break;
			case 'p':
				/* proxy server parsing server:port */
				tmp=strrchr(optarg,':');
				bench_req.proxyhost=optarg;
				if(tmp==NULL){
					break;
				}
				if(tmp==optarg){
					fprintf(stderr,"Error in option --proxy %s: Missing hostname.\n",optarg);
					return 2;
				}
				if(tmp==optarg+strlen(optarg)-1){
					fprintf(stderr,"Error in option --proxy %s Port number is missing.\n",optarg);
					return 2;
				}
				*tmp='\0'; // 截断proxyhost和proxyport
				bench_req.proxyport=atoi(tmp+1);
				printf("\nproxy host: %s, proxy port: %d\n", bench_req.proxyhost, bench_req.proxyport );
				break;
			case ':':
			case 'h':
			case '?':
				usage();
				return 2;
				break;// 为了语法通过，加的break,其实根本没用
		}
	}

	if(optind==argc) {
		fprintf(stderr,"webbench: Missing URL!\n");
		usage();
		printf("usage ");
		return 2;
	}

	if(bench_req.clients==0) bench_req.clients=1;
	if(bench_req.benchtime==0) bench_req.benchtime=30;

	/* Copyright */
	fprintf(stderr, "Webbench - Simple Web Benchmark %s %s \n", PROGRAM_VERSION,
			"Copyright (c) Radim Kolar 1997-2004, GPL Open Source Software.\n"
	);
	bench_req.url = argv[optind];
    build_request(&bench_req); // 构建请求体
	f1("abcd");

	printf("Runing info: ");

	if(bench_req.clients==1)
		printf("1 client");
	else
		printf("%d clients",bench_req.clients);

	printf(", running %d sec", bench_req.benchtime);

	if(bench_req.force) printf(", early socket close");
	if(bench_req.proxyhost!=NULL) printf(", via proxy server %s:%d",bench_req.proxyhost,bench_req.proxyport);
	if(bench_req.force_reload) printf(", forcing reload");

	printf(".\n");
//	return 0;
    return bench(&bench_req);
}


void usage(void)
{
	fprintf(stderr,
			"webbench [option]... URL\n"
			"  -f|--force               Don't wait for reply from server.\n"
			"  -r|--reload              Send reload request - Pragma: no-cache.\n"
			"  -t|--time <sec>          Run benchmark for <sec> seconds. Default 30.\n"
			"  -p|--proxy <server:port> Use proxy server for request.\n"
			"  -c|--clients <n>         Run <n> HTTP clients at once. Default one.\n"
			"  -9|--http09              Use HTTP/0.9 style requests.\n"
			"  -1|--http10              Use HTTP/1.0 protocol.\n"
			"  -2|--http11              Use HTTP/1.1 protocol.\n"
			"  --get                    Use GET request method.\n"
			"  --head                   Use HEAD request method.\n"
			"  --options                Use OPTIONS request method.\n"
			"  --trace                  Use TRACE request method.\n"
			"  -?|-h|--help             This information.\n"
			"  -V|--version             Display program version.\n"
	);
}



socket.h
/* $Id: socket.c 1.1 1995/01/01 07:11:14 cthuang Exp $
 *
 * This module has been modified by Radim Kolar for OS/2 emx
 */

/***********************************************************************
  module:       socket.c
  program:      popclient
  SCCS ID:      @(#)socket.c    1.5  4/1/94
  programmer:   Virginia Tech Computing Center
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3
  description:  UNIX sockets code.
 ***********************************************************************/


#ifndef CPP_LEARN_A_H
#define CPP_LEARN_A_H

#include <cstdlib>


extern char* PROGRAM_VERSION;
#define METHOD_GET 0
#define METHOD_HEAD 1
#define METHOD_OPTIONS 2
#define METHOD_TRACE 3

struct bench_request{
	int method;
	char* url; // 用户传进来的url
	int http10; // http版本，根据请求方法来定
	char* proxyhost;
	int proxyport; // 接受的代理服务器
	char* host;
	char *request;// 生成的请求报文

	int clients;
	int benchtime;

	int force;
	int force_reload;

	bench_request(){
		http10=1;
		method=METHOD_GET;
		clients=1; // 这是一个整数
		force=0;
		force_reload=0;
		proxyport=80;
		proxyhost=NULL;
		host= NULL ;
		benchtime=30; // 默认执行30秒
		request= NULL ;

	}
};

int Socket(const char *host, int clientPort);

#endif

bench.cpp
/*
* (C) Radim Kolar 1997-2004
* This is free software, see GNU Public License version 2 for
* details.
*
* Simple forking WWW Server benchmark:
*
* Usage:
*   webbench --help
*
* Return codes:
*    0 - sucess
*    1 - benchmark failed (server is not on-line)
*    2 - bad param
*    3 - internal error, fork failed
* 
*/ 

#include <unistd.h>
#include <sys/param.h>
#include <rpc/types.h>
#include <getopt.h>
#include <strings.h>
#include <time.h>
#include <signal.h>
#include <string>
#include <iostream>
#include <sys/socket.h>

#include "socket.h" // 没有看错，是include

/* globals */
// http的版本 @http
/* 0 - http/0.9, 1 - http/1.0, 2 - http/1.1
 * 问题是： 这版本之间有什么区别呢？
 *
 * */

/* internal */

volatile int timerexpired=0;
int speed=0;
int failed=0;
int bytes=0;
int mypipe[2];

void benchcore(char* host, int port , bench_request* benchreq);
static void alarm_handler(int signal)
{
	timerexpired=1;
}


int bench(bench_request* req){
    int i,j,k;
    pid_t pid=0;
    FILE *f;
	int clients = req->clients;
    char* host = req->host;
	char* proxyhost = req->proxyhost;
	int proxyport = req->proxyport;
	char* request = req->request;
	int benchtime = req->benchtime;
	int force = req->force;

    /* check avaibility of target server */
	// step（1）: 首先测试TCP连接是否合法；
    i=Socket(proxyhost==NULL?host:proxyhost,proxyport);
    if(i<0) {
        fprintf(stderr,"\nConnect to server failed. Aborting benchmark.\n");
        return 1;
    }
    close(i);
	// step2 (2）:调用pipe建立管道；
    /* create pipe */
    if(pipe(mypipe))
    {
        perror("pipe failed.");
        return 3;
    }

    /* fork childs */
    for(i=0;i<clients;i++){
        pid=fork();
        if(pid <= (pid_t) 0) // TODO
        {
            /* child process or error*/
            sleep(1); /* make childs faster */
            break;
        }
    }

    if( pid < (pid_t) 0){
        fprintf(stderr,"problems forking worker no. %d\n",i);
        perror("fork failed.");
        return 3;
    }

    if(pid == (pid_t) 0){
        /* I am a child */
        if(proxyhost==NULL) // 启用代理
            benchcore(host,proxyport,req); // 把request发给代理或者目标主机， request本身就带host
        else
            benchcore(proxyhost,proxyport,req);

        /* write results to pipe */
        f=fdopen(mypipe[1],"w");
        if(f==NULL){
            perror("open pipe for writing failed.");
            return 3;
        }
        /* fprintf(stderr,"Child - %d %d\n",speed,failed); */
        fprintf(f,"%d %d %d\n",speed,failed,bytes);
        fclose(f);
        return 0;
    }
    else
    {
        f=fdopen(mypipe[0],"r");
        if(f==NULL)
        {
            perror("open pipe for reading failed.");
            return 3;
        }

        setvbuf(f,NULL,_IONBF,0);

        speed=0;
        failed=0;
        bytes=0;

        while(1)
        {
            pid=fscanf(f,"%d %d %d",&i,&j,&k); // 从管道中读入3个数字
            if(pid<2) // 只有至多一个进程了
            {
                fprintf(stderr,"Some of our childrens died.\n");
                break;
            }

            speed+=i;
            failed+=j;
            bytes+=k;

            /* fprintf(stderr,"*Knock* %d %d read=%d\n",speed,failed,pid); */
            if(--clients==0) break;
        }

        fclose(f);

        printf("\nSpeed=%d pages/min, %d bytes/sec.\nRequests: %d susceed, %d failed.\n",
            (int)((speed+failed)/(benchtime/60.0f)),
            (int)(bytes/(float)benchtime),
            speed,
            failed);
    }

    return i;
}

void benchcore(char* host, int port , bench_request* benchreq){
    int rlen;
    char buf[1500];
    int s,i;
    struct sigaction sa;

    int force = benchreq->force;
	char *req =  benchreq->request;
	int benchtime = benchreq->benchtime;
	int http10 = benchreq->http10;

    /* setup alarm signal handler */
    sa.sa_handler=alarm_handler;
    sa.sa_flags=0;
    if(sigaction(SIGALRM,&sa,NULL))
        exit(3);

    alarm(benchtime); // after benchtime,then exit

    rlen=strlen(req);
    nexttry:while(1){
        if(timerexpired){
            if(failed>0){
                /* fprintf(stderr,"Correcting failed by signal\n"); */
                failed--;
            }
            return;
        }

        s=Socket(host,port);
        if(s<0) {failed++;continue;}
        if(rlen!=write(s,req,rlen)) {failed++;close(s);continue;}
        if(http10==0)
        if(shutdown(s,1)) { failed++;close(s);continue;}

        if(force==0){ /* read all available data from socket */
            while(1){
                if(timerexpired) break;
                i=read(s,buf,1500);
				/* fprintf(stderr,"%d\n",i); */
				if(i<0) {
					failed++;
					close(s);
					goto nexttry;
				} else if(i==0) break;
				else{
					bytes+=i;
				}
            }
        }

        if(close(s)) { /* 连接提前断开，这里有点看不懂 */
        	failed++;
        	continue;
        }
        speed++;
    }// nexttry end
}



socket2.h
/* $Id: socket.c 1.1 1995/01/01 07:11:14 cthuang Exp $
 *
 * This module has been modified by Radim Kolar for OS/2 emx
 */
/***********************************************************************
  module:       socket.c
  program:      popclient
  SCCS ID:      @(#)socket.c    1.5  4/1/94
  programmer:   Virginia Tech Computing Center
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  UNIX sockets code.
 ***********************************************************************/

#include <sys/types.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/time.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>



int Socket(const char *host, int clientPort)
{
    int sock;
    unsigned long inaddr;
    struct sockaddr_in ad;
    struct hostent *hp;

    memset(&ad, 0, sizeof(ad));
    ad.sin_family = AF_INET;

    inaddr = inet_addr(host); // 将一个点分十进制的IP字符串转化为网络字节序二进制位对应的无符号整型。
    if (inaddr != INADDR_NONE)
        memcpy(&ad.sin_addr, &inaddr, sizeof(inaddr));
    else
    {
        hp = gethostbyname(host);
        if (hp == NULL)
            return -1;
        memcpy(&ad.sin_addr, hp->h_addr, hp->h_length);
    }
    ad.sin_port = htons(clientPort);

    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0)
        return sock;
    if (connect(sock, (struct sockaddr *)&ad, sizeof(ad)) < 0)
        return -1;
    return sock;
}


socket.cpp
/* $Id: socket.c 1.1 1995/01/01 07:11:14 cthuang Exp $
 *
 * This module has been modified by Radim Kolar for OS/2 emx
 */
/***********************************************************************
  module:       socket.c
  program:      popclient
  SCCS ID:      @(#)socket.c    1.5  4/1/94
  programmer:   Virginia Tech Computing Center
  compiler:     DEC RISC C compiler (Ultrix 4.1)
  environment:  DEC Ultrix 4.3 
  description:  UNIX sockets code.
 ***********************************************************************/

#include <sys/types.h>
#include <sys/socket.h>
#include <fcntl.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <netdb.h>
#include <sys/time.h>
#include <string.h>
#include <unistd.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>


#define METHOD_GET 0
#define METHOD_HEAD 1
#define METHOD_OPTIONS 2
#define METHOD_TRACE 3

char* PROGRAM_VERSION = "1.5";

int Socket(const char *host, int clientPort)
{
    int sock;
    unsigned long inaddr;
    struct sockaddr_in ad;
    struct hostent *hp;

    memset(&ad, 0, sizeof(ad));
    ad.sin_family = AF_INET;

    inaddr = inet_addr(host); // 将一个点分十进制的IP字符串转化为网络字节序二进制位对应的无符号整型。
    if (inaddr != INADDR_NONE)
        memcpy(&ad.sin_addr, &inaddr, sizeof(inaddr));
    else
    {
        hp = gethostbyname(host);
        if (hp == NULL)
            return -1;
        memcpy(&ad.sin_addr, hp->h_addr, hp->h_length);
    }
    ad.sin_port = htons(clientPort);

    sock = socket(AF_INET, SOCK_STREAM, 0);
    if (sock < 0)
        return sock;
    if (connect(sock, (struct sockaddr *)&ad, sizeof(ad)) < 0)
        return -1;
    return sock;
}


webbench.c
/*
* (C) Radim Kolar 1997-2004
* This is free software, see GNU Public License version 2 for
* details.
*
* Simple forking WWW Server benchmark:
*
* Usage:
*   webbench --help
*
* Return codes:
*    0 - sucess
*    1 - benchmark failed (server is not on-line)
*    2 - bad param
*    3 - internal error, fork failed
* 
*/ 

#include "socket.h" // 没有看错，是include
#include <unistd.h>
#include <sys/param.h>
#include <rpc/types.h>
#include <getopt.h>
#include <strings.h>
#include <time.h>
#include <signal.h>
#include <thread>

/* values */
volatile int timerexpired=0;
int speed=0;
int failed=0;
int bytes=0;

/* globals */
// http的版本 @http
/* 0 - http/0.9, 1 - http/1.0, 2 - http/1.1
 * 问题是： 这版本之间有什么区别呢？
 *
 * */
int http10=1;

/* Allow: GET, HEAD, OPTIONS, TRACE */
#define METHOD_GET 0
#define METHOD_HEAD 1
#define METHOD_OPTIONS 2
#define METHOD_TRACE 3
#define PROGRAM_VERSION "1.5"

int method=METHOD_GET;
int clients=1; // 这是一个整数
int force=0;
int force_reload=0;
int proxyport=80;
char *proxyhost=NULL;
int benchtime=30; // 默认执行30秒

/* internal */
int mypipe[2];
char host[MAXHOSTNAMELEN];
#define REQUEST_SIZE 2048 // 一个请求 2K 字节啊
char request[REQUEST_SIZE]; // 这是一个字符串数组啊，神奇，也就是一个字符串咯

static const struct option long_options[]=
{
    {"force",no_argument,&force,1},
    {"reload",no_argument,&force_reload,1},
    {"time",required_argument,NULL,'t'},
    {"help",no_argument,NULL,'?'},
    {"http09",no_argument,NULL,'9'},
    {"http10",no_argument,NULL,'1'},
    {"http11",no_argument,NULL,'2'},
    {"get",no_argument,&method,METHOD_GET},
    {"head",no_argument,&method,METHOD_HEAD},
    {"options",no_argument,&method,METHOD_OPTIONS},
    {"trace",no_argument,&method,METHOD_TRACE},
    {"version",no_argument,NULL,'V'},
    {"proxy",required_argument,NULL,'p'},
    {"clients",required_argument,NULL,'c'},
    {NULL,0,NULL,0}
};

/* prototypes */
static void benchcore(const char* host,const int port, const char *request);
static int bench(void);
static void build_request(const char *url);

static void alarm_handler(int signal)
{
    timerexpired=1;
}	

static void usage(void)
{
    fprintf(stderr,
            "webbench [option]... URL\n"
            "  -f|--force               Don't wait for reply from server.\n"
            "  -r|--reload              Send reload request - Pragma: no-cache.\n"
            "  -t|--time <sec>          Run benchmark for <sec> seconds. Default 30.\n"
            "  -p|--proxy <server:port> Use proxy server for request.\n"
            "  -c|--clients <n>         Run <n> HTTP clients at once. Default one.\n"
            "  -9|--http09              Use HTTP/0.9 style requests.\n"
            "  -1|--http10              Use HTTP/1.0 protocol.\n"
            "  -2|--http11              Use HTTP/1.1 protocol.\n"
            "  --get                    Use GET request method.\n"
            "  --head                   Use HEAD request method.\n"
            "  --options                Use OPTIONS request method.\n"
            "  --trace                  Use TRACE request method.\n"
            "  -?|-h|--help             This information.\n"
            "  -V|--version             Display program version.\n"
           );
}

int main(int argc, char *argv[])
{
    int opt=0;
    int options_index=0;
    char *tmp=NULL;

    if(argc==1)
    {
        usage();
        return 2;
    } 

    while((opt=getopt_long(argc,argv,"912Vfrt:p:c:?h",long_options,&options_index))!=EOF )
    {
        switch(opt)
        {
            case  0 : break;
            case 'f': force=1;break;
            case 'r': force_reload=1;break; 
            case '9': http10=0;break;
            case '1': http10=1;break;
            case '2': http10=2;break;
            case 'V': printf(PROGRAM_VERSION"\n");exit(0);
            case 't': benchtime=atoi(optarg);break;	     
            case 'p': 
            /* proxy server parsing server:port */
            tmp=strrchr(optarg,':');
            proxyhost=optarg;
            if(tmp==NULL)
            {
                break;
            }
            if(tmp==optarg)
            {
                fprintf(stderr,"Error in option --proxy %s: Missing hostname.\n",optarg);
                return 2;
            }
            if(tmp==optarg+strlen(optarg)-1)
            {
                fprintf(stderr,"Error in option --proxy %s Port number is missing.\n",optarg);
                return 2;
            }
            *tmp='\0';
            proxyport=atoi(tmp+1);break;
            case ':':
            case 'h':
            case '?': usage();return 2;break;
            case 'c': clients=atoi(optarg);break;
        }
    }

    if(optind==argc) {
        fprintf(stderr,"webbench: Missing URL!\n");
        usage();
        return 2;
    }

    if(clients==0) clients=1;
    if(benchtime==0) benchtime=30;
 
    /* Copyright */
    fprintf(stderr,"Webbench - Simple Web Benchmark "PROGRAM_VERSION"\n"
            "Copyright (c) Radim Kolar 1997-2004, GPL Open Source Software.\n"
            );
 
    build_request(argv[optind]);
 
    // print request info ,do it in function build_request
    /*printf("Benchmarking: ");
 
    switch(method)
    {
        case METHOD_GET:
        default:
        printf("GET");break;
        case METHOD_OPTIONS:
        printf("OPTIONS");break;
        case METHOD_HEAD:
        printf("HEAD");break;
        case METHOD_TRACE:
        printf("TRACE");break;
    }
    
    printf(" %s",argv[optind]);
    
    switch(http10)
    {
        case 0: printf(" (using HTTP/0.9)");break;
        case 2: printf(" (using HTTP/1.1)");break;
    }
 
    printf("\n");
    */

    printf("Runing info: ");

    if(clients==1) 
        printf("1 client");
    else
        printf("%d clients",clients);

    printf(", running %d sec", benchtime);
    
    if(force) printf(", early socket close");
    if(proxyhost!=NULL) printf(", via proxy server %s:%d",proxyhost,proxyport);
    if(force_reload) printf(", forcing reload");
    
    printf(".\n");
    
    return bench();
}

//@b
// 根据参数行命令建立一个Http请求，并将http请求放入到全局数据request中。
void build_request(const char *url)
{
    char tmp[10];
    int i;
    //bzero(host,MAXHOSTNAMELEN);
    //bzero(request,REQUEST_SIZE);
    memset(host,0,MAXHOSTNAMELEN);
    memset(request,0,REQUEST_SIZE);
    if(force_reload && proxyhost!=NULL && http10<1) http10=1;
    if(method==METHOD_HEAD && http10<1) http10=1;
    if(method==METHOD_OPTIONS && http10<2) http10=2;
    if(method==METHOD_TRACE && http10<2) http10=2;

    /*
     *
     * */

    switch(method)
    {
        default:
        case METHOD_GET: strcpy(request,"GET");break;
        case METHOD_HEAD: strcpy(request,"HEAD");break;
        case METHOD_OPTIONS: strcpy(request,"OPTIONS");break;
        case METHOD_TRACE: strcpy(request,"TRACE");break;
    }

    strcat(request," ");

    if(NULL==strstr(url,"://"))
    {
        fprintf(stderr, "\n%s: is not a valid URL.\n",url);
        exit(2);
    }
    if(strlen(url)>1500)
    {
        fprintf(stderr,"URL is too long.\n");
        exit(2);
    }
    if (0!=strncasecmp("http://",url,7)) 
    { 
        fprintf(stderr,"\nOnly HTTP protocol is directly supported, set --proxy for others.\n");
        exit(2);
    }
    
    /* protocol/host delimiter */
    i=strstr(url,"://")-url+3; //举个例子， http://www.baidu.com, 现在等于www的下标
    printf("%d\n",i);


    printf("strchr: %s\n", strchr(url+i,'/'));
    printf("hahah\n");
    if(strchr(url+i,'/')==NULL) {
        fprintf(stderr,"\nInvalid URL syntax - hostname don't ends with '/'.\n");
        exit(2);
    }

    printf("@proxy %s\n", proxyhost);
    
    if(proxyhost==NULL)
    {
    	printf("@1: %s\n", index(url+i,':'));
		printf("@2: %s\n", index(url+i,'/'));
        /* get port from hostname */
        if(index(url+i,':')!=NULL && index(url+i,':')<index(url+i,'/')) // i就是 baidu.com的下标
        {
            strncpy(host,url+i,strchr(url+i,':')-url-i);
            //bzero(tmp,10);
            memset(tmp,0,10);
            strncpy(tmp,index(url+i,':')+1,strchr(url+i,'/')-index(url+i,':')-1);
            /* printf("tmp=%s\n",tmp); */
			printf("the @port 1: %s ", tmp);
            proxyport=atoi(tmp);
            printf("the @port %d ", proxyport);
            if(proxyport==0) proxyport=80;
        } 
        else
        {
            strncpy(host,url+i,strcspn(url+i,"/"));
        }
        // printf("Host=%s\n",host);
        strcat(request+strlen(request),url+i+strcspn(url+i,"/"));
    } 
    else
    {
        // printf("ProxyHost=%s\nProxyPort=%d\n",proxyhost,proxyport);
        strcat(request,url);
    }

    if(http10==1)
        strcat(request," HTTP/1.0");
    else if (http10==2)
        strcat(request," HTTP/1.1");
  
    strcat(request,"\r\n"); // 拼接换行符
  
    if(http10>0)
        strcat(request,"User-Agent: WebBench "PROGRAM_VERSION"\r\n");

    if(proxyhost==NULL && http10>0)
    {
        strcat(request,"Host: ");
        strcat(request,host);
        strcat(request,"\r\n");
    }
 
    if(force_reload && proxyhost!=NULL)
    {
        strcat(request,"Pragma: no-cache\r\n");
    }
  
    if(http10>1)
        strcat(request,"Connection: close\r\n");
    
    /* add empty line at end */
    if(http10>0) strcat(request,"\r\n"); 
    
    printf("\nRequest:\n%s\n",request);
}

/* vraci system rc error kod */
static int bench(void)
{
    int i,j,k;	
    pid_t pid=0;
    FILE *f;

    /* check avaibility of target server */
	// step（1）: 首先测试TCP连接是否合法；
    i=Socket(proxyhost==NULL?host:proxyhost,proxyport);
    if(i<0) { 
        fprintf(stderr,"\nConnect to server failed. Aborting benchmark.\n");
        return 1;
    }
    close(i);
	// step2 (2）:调用pipe建立管道；
    /* create pipe */
    if(pipe(mypipe))
    {
        perror("pipe failed.");
        return 3;
    }

    /* not needed, since we have alarm() in childrens */
    /* wait 4 next system clock tick */
    /*
    cas=time(NULL);
    while(time(NULL)==cas)
    sched_yield();
    */

    /* fork childs */
    for(i=0;i<clients;i++)
    {
        pid=fork();
        if(pid <= (pid_t) 0) // TODO
        {
            /* child process or error*/
            sleep(1); /* make childs faster */
            break;
        }
    }

    if( pid < (pid_t) 0)
    {
        fprintf(stderr,"problems forking worker no. %d\n",i);
        perror("fork failed.");
        return 3;
    }

    if(pid == (pid_t) 0)
    {
        /* I am a child */
        if(proxyhost==NULL)
            benchcore(host,proxyport,request);
        else
            benchcore(proxyhost,proxyport,request);

        /* write results to pipe */
        f=fdopen(mypipe[1],"w");
        if(f==NULL){
            perror("open pipe for writing failed.");
            return 3;
        }
        /* fprintf(stderr,"Child - %d %d\n",speed,failed); */
        fprintf(f,"%d %d %d\n",speed,failed,bytes);
        fclose(f);

        return 0;
    } 
    else
    {
        f=fdopen(mypipe[0],"r");
        if(f==NULL) 
        {
            perror("open pipe for reading failed.");
            return 3;
        }
        
        setvbuf(f,NULL,_IONBF,0);
        
        speed=0;
        failed=0;
        bytes=0;
    
        while(1)
        {
            pid=fscanf(f,"%d %d %d",&i,&j,&k);
            if(pid<2) // 只有至多一个进程了
            {
                fprintf(stderr,"Some of our childrens died.\n");
                break;
            }
            
            speed+=i;
            failed+=j;
            bytes+=k;
        
            /* fprintf(stderr,"*Knock* %d %d read=%d\n",speed,failed,pid); */
            if(--clients==0) break;
        }
    
        fclose(f);

        printf("\nSpeed=%d pages/min, %d bytes/sec.\nRequests: %d susceed, %d failed.\n",
            (int)((speed+failed)/(benchtime/60.0f)),
            (int)(bytes/(float)benchtime),
            speed,
            failed);
    }
    
    return i;
}

//@b2
void benchcore(const char *host,const int port,const char *req)
{
    int rlen;
    char buf[1500];
    int s,i;
    struct sigaction sa;

    /* setup alarm signal handler */
    sa.sa_handler=alarm_handler;
    sa.sa_flags=0;
    if(sigaction(SIGALRM,&sa,NULL))
        exit(3);
    
    alarm(benchtime); // after benchtime,then exit

    rlen=strlen(req);
    nexttry:while(1){
        if(timerexpired){
            if(failed>0){
                /* fprintf(stderr,"Correcting failed by signal\n"); */
                failed--;
            }
            return;
        }
        
        s=Socket(host,port);                          
        if(s<0) {failed++;continue;}
        if(rlen!=write(s,req,rlen)) {failed++;close(s);continue;}
        if(http10==0) 
        if(shutdown(s,1)) { failed++;close(s);continue;}

        if(force==0){ /* read all available data from socket */
            while(1){
                if(timerexpired) break; 
                i=read(s,buf,1500);
				/* fprintf(stderr,"%d\n",i); */
				if(i<0) {
					failed++;
					close(s);
					goto nexttry;
				} else if(i==0) break;
				else{
					bytes+=i;
				}
            }
        }

        if(close(s)) { /* 连接提前断开，这里有点看不懂 */
        	failed++;
        	continue;
        }
        speed++;
    }// nexttry end
}
